src/modem.rs

```diff
@@ -462,39 +597,45 @@ pub mod sim {
             }
         }
 
-        pub fn get_signal_quality<IO: embedded_svc::io::Read + embedded_svc::io::Write>(
-            comm: &mut IO,
-        ) -> Result<(i32, i32), ModemError> {
-            let mut buff = [0u8; 64];
-            let cmd = CommandBuilder::create_execute(&mut buff, true)
+        pub fn get_signal_quality(
+            comm: &mut UartDriver,
+            buff: &mut [u8],
+        ) -> Result<(RSSI, BitErrorRate), ModemError> {
+            let cmd = CommandBuilder::create_execute(buff, true)
                 .named("+CSQ")
                 .finish()?;
 
             comm.write(cmd).map_err(|_| ModemError::IO)?;
 
-            let len = comm.read(&mut buff).map_err(|_| ModemError::IO)?;
+            let len = comm
```

Comment:
I don't think this type of reading (and all others in all `get_*` methods below and above) is reliable.

You are not taking into account, _that you might not get the complete response as a single shot to `read`_. Here for example, the fact that you get `\r\n+CSQ: 19,99\r\n\r\nOK\r\n` out of a single `read` is a pure coincidence, and might break any time. Just think what would've happened, if I've kept the other code in `UartDriver::read`, which did return - on an empty ringbuffer - **a single byte** after waiting for it. Havoc.

I think you need a utility method that probably takes a callback closure of sorts, where the callback closure is containing the parsing code of `CommandParse::parse` and then - in the absence of a reliable end-of-command-response termination character (like a CR or LF) that you might wait for - is trying to apply parsing on the buffer repeatedly, and if the data in the buffer is incomplete - tries to read a bit more at the end of the buffer.

For that purpose, you might want to take a look at `embedded_io_async::BufRead` as its API is handy for such cases (and implementing `BufRead` over `Read` is trivial, if not done already in the lib itself.

Alternative: you might want to layer a async fn which reads line-by-line from a `BufRead` and refactor your parsing to line-by-line (i.e. CommandParser would expect lines somehow).

Reading line-by-line simply means you call `BufRead::read` repeatedly until you notice a sequence of `\r\n` in the buffer, then consume everything until (and trimming) `\r\n` and pass this as the "line" to your stateful parser of sorts.

Reply:
A third alternative (but I find this really ugly) and it will make the negotiation slower, is to `read` in your buffer - in a loop - until a given timeout. Where the expectation is that until this timeout had expired, the modem would've replied - in your buffer - with the complete response to your command.

But of course this introduces an arbitrary variable in all of this (the "read till you get the whole response" timeout, which would be - what? - a trial and error exercise).

Reply:
Looking at the `atat` crate, [here is a similar "loop" to what I mean](https://github.com/FactbirdHQ/atat/blob/master/atat/src/asynch/simple_client.rs#L52). Basically, the loop reads at the end of the buffer and then hands the buffer to a "digester". The digester would return `DigestResult::None` if the data in the buffer is "incomplete" (i.e. we have to read more to get the whole response).

Reply:
> I think you need a utility method that probably takes a callback closure of sorts, where the callback closure is containing the parsing code of CommandParse::parse and then - in the absence of a reliable end-of-command-response termination character (like a CR or LF) that you might wait for - is trying to apply parsing on the buffer repeatedly, and if the data in the buffer is incomplete - tries to read a bit more at the end of the buffer.

I do think at least that `\r\nOK\r\n` is a reliable end of command termination.

Reply:
This is what I came up with:
```

pub enum DigestError {
    Error,
    Continue,
    BufferOverflow,
}

pub trait Digester {
    /// Termination characters if the response indicates the OK state.
    const OK_TERMINATOR: &'static str;
    /// Termination characters if the response indicates the Error state.
    const ERR_TERMINATOR: &'static str;
    /// parse the buffer, searching for finaliser.
    /// If the
    fn digest(&mut self, data: &[u8]) -> Result<&[u8], DigestError>;
}

pub struct SimDigester<const N: usize> {
    buf: [u8; N],
    pos: usize,
}

impl<const N: usize> Digester for SimDigester<N> {
    const OK_TERMINATOR: &'static str = "\r\nOK\r\n";
    const ERR_TERMINATOR: &'static str = "\r\nERROR\r\n";

    fn digest(&mut self, data: &[u8]) -> Result<&[u8], DigestError> {
        let incoming_len = data.len();
        if self.pos + incoming_len > N {
            return Err(DigestError::BufferOverflow);
        }
        self.buf[self.pos..self.pos + incoming_len].copy_from_slice(data);
        self.pos += incoming_len;
        // look for the terminator, get to the choppa
        let ok_len = Self::OK_TERMINATOR.len();
        let end = &self.buf[(self.pos - ok_len)..self.pos];
        if end == Self::OK_TERMINATOR.as_bytes() {
            // ready to parse
            return Ok(&self.buf[..self.pos]);
        }
        let err_len = Self::ERR_TERMINATOR.len();
        let end: &[u8] = &self.buf[(self.pos - err_len)..self.pos];
        if end == Self::ERR_TERMINATOR.as_bytes() {
            // result indicated failure
            return Err(DigestError::Error);
        }
        Err(DigestError::Continue)
    }
}
```

I can also modify `Digester` to be more like `BufRead`, if that is better approach?

Reply:
I don't think you need the digester thing given that you can just use `BufRead` indeed.

Be mindful of the last response to the command with which you switch to PPP. There you might have to read byte-by-byte, or else you might read a bit into the first PPP packet right after the end of the last AT response. Speculating of course, not sure if the modem will start sending PPP packets immediately, but you might need to check that.

Reply:
I think it might be good while I'm at it to introduce `BufReader` struct to `embedded-io`, similar to [std](https://doc.rust-lang.org/std/index.html)::[io](https://doc.rust-lang.org/std/io/index.html)::[BufReader](https://doc.rust-lang.org/std/io/struct.BufReader.html#).

```

pub struct BufReader<const N: usize, R: ?Sized> {
    buf: [u8; N],
    pos: usize,
    inner: R,
}

impl<const N: usize, R: Read> ErrorType for BufReader<N, R> {
    type Error = R::Error;
}

impl<const N: usize, R: Read> BufRead for BufReader<N, R> {
    fn consume(&mut self, amt: usize) {
        // self
        self.pos += amt;
    }

    fn fill_buf(&mut self) -> Result<&[u8], Self::Error> {
        // fill the inner buffer
        let read_count = self.inner.read(&mut self.buf[self.pos..])?;
        // return the read bytes
        Ok(&self.buf[self.pos..(self.pos + read_count)])
    }
}
```

Reply:
I struggled to get this to work with `embedded_io::BufRead`, and even more struggle with `std::io::BufReader`.

However, this is my current approach, let me know if this is suitable (currently targeting `std::io::Read`, so that I can debug it on my local PC)

```rust

pub struct SIM7600(CommunicationMode);

impl<T: std::io::Read + std::io::Write> SimModem<T> for SIM7600 {
    fn negotiate(&mut self, comm: &mut T, buffer: &mut [u8]) -> Result<(), ModemError> {
        reset(comm, buffer)?;
// others ignored here
        self.0 = CommunicationMode::Data;
        Ok(())
    }

    fn get_mode(&self) -> &CommunicationMode {
        &self.0
    }
}



fn reset<T: std::io::Read + std::io::Write>(
    comm: &mut T,
    buff: &mut [u8],
) -> Result<(), ModemError> {
    let cmd = CommandBuilder::create_execute(buff, false)
        .named("ATZ0")
        .finish()?;
    log::info!("Send Reset");

    comm.write(cmd).map_err(|_| ModemError::IO)?;

    // read until terminator
    let len = read_until_term(comm, buff)?;

    log::info!("got response{:?}", core::str::from_utf8(&buff[..len]));
    CommandParser::parse(&buff[..len])
        .expect_identifier(b"ATZ0\r")
        .expect_identifier(b"\r\nOK\r\n")
        .finish()?;
    Ok(())
}

const OK_TERMINATOR: &'static str = "\r\nOK\r\n";
const ERR_TERMINATOR: &'static str = "\r\nERR\r\n";

fn read_until_term<T: std::io::Read>(comm: &mut T, buff: &mut [u8]) -> Result<usize, ModemError> {
    let mut pos = 0;

    loop {
        // start filling the buffer
        let len = comm.read(&mut buff[pos..]).map_err(|w| ModemError::IO)?;
        if len == 0 {
            continue;
        }
        pos += len;
        log::info!(
            "Len: {}, data = {:?}",
            len,
            std::str::from_utf8(&buff[..pos])
        );

        let ok_len = OK_TERMINATOR.len();
        if pos >= ok_len {
            let end = &buff[pos - ok_len..pos];
            log::info!(" end = {:?}", std::str::from_utf8(&buff[pos - ok_len..pos]));
            if end == OK_TERMINATOR.as_bytes() {
                // buff[..pos].copy_from_slice(data);
                return Ok(pos);
            }
        }

        let err_len = ERR_TERMINATOR.len();
        if pos >= err_len {
            let end = &buff[pos - err_len..pos];

            if end == ERR_TERMINATOR.as_bytes() {
                return Err(ModemError::DigestError);
            }
        }
        // return Err(ModemError::DigestError);
    }
}
```

Reply:
Also this:
```rust
fn fill_buf(&mut self) -> Result<&[u8], Self::Error> {
        // fill the inner buffer
        let read_count = self.inner.read(&mut self.buf[self.pos..])?;
        // return the read bytes
        Ok(&self.buf[self.pos..(self.pos + read_count)])
}
```

needs to be:
```rust
fn fill_buf(&mut self) -> Result<&[u8], Self::Error> {
        // fill the inner buffer
        let read_count = self.inner.read(&mut self.buf[self.pos..])?;
        self.pos += read_count; // You need to update the end of your data
        // return the read bytes
        Ok(&self.buf[..self.pos]) // You always return ALL of your data UNTIL the user - via `consume` - tells you to remove N bytes from the beginning of the buffer
}
```

Reply:
As for the approach that uses a raw `Read` (`read_until_term`)... I'm not sure it is correct either. The problem is, the call to `T::read` might read data **beyond** your line terminator (i.e. data from the _next_ line). You need to preserve this data somewhere. And then re-use it in the _next_ call to `read_until_term`.

That's why STD line readers are implemented with `BufRead` rather than with `Read`. Since `BufRead` "owns" the buffer, it can **preserve** whatever remains in it (i.e. whatever is still not declared as "consumed" by the downstream code) for your _next_ call to `read_until_terminator`, that would use `BufRead`.

Reply:
(Update: remove the comment above `Line::Continue` as you actually can't get rid of it as the `BufRead::fetch_buf` does not guarantee you to fetch extra, if there is already content in the buffer. The only way to recover whole lines from the above example code is by appending continue+continue+...+end in a large enough buffer. Or by switching your parsing logic to an iterator-like approach where you process byte-by-byte (= char-by-char for ascii).

Reply:
> Be mindful of the last response to the command with which you switch to PPP. There you might have to read byte-by-byte, or else you might read a bit into the first PPP packet right after the end of the last AT response. Speculating of course, not sure if the modem will start sending PPP packets immediately, but you might need to check that.

With this requirement, I think we will need a byte-by-byte approach anyway, at least for the last packet. 

> The above is wrong.
> consume needs to physically remove amt bytes from the beginning of the buffer (i.e. the bytes of &buf[..amt]), by doing an overlapping-copy of &buf[amt..pos] (src) into &buf[..pos - amt] (dest).
> And then you need to do self.pos -= amt;.

I think I fundamentally didn't understand `BufRead`, but I think I get it now.

> Why? UartDriver already does implement Read, so BufReader<N, UartDriver<'d>> would just work.

Yes you are right. I have implemented a `BufReader` targeting `embedded-io`. https://github.com/rust-embedded/embedded-hal/pull/627, that should work with `UartDriver`. 

> As for the approach that uses a raw Read (read_until_term)... I'm not sure it is correct either. The problem is, the call to T::read might read data beyond your line terminator (i.e. data from the next line). You need to preserve this data somewhere. And then re-use it in the next call to read_until_term.

In this case, i'm not sure that is so much of an issue because the AT commands are request ->response, although I see some mention of the idea of "URC" or Unsolicited Result Code, but I haven't encountered them yet, at least for SIM7600.

> Not sure what is the difficulty with BufRead but here's something which returns stuff line by line (with a simpler terminator indeed, but you can extend to a more complex one):

BufRead seemed to be returning the entire `buffer[self.pos..]`, including the un-initialized bits. I think I still prefer an approach that reads the entire buffer until `\r\nOK\r\n` is found, but I am still working on it.

Reply:
> BufRead seemed to be returning the entire `buffer[self.pos..]`, including the un-initialized bits. I think I still prefer an approach that reads the entire buffer until `\r\nOK\r\n` is found, but I am still working on it.

No, `BufRead::fill_buf` should - by definition -  **only** return a slice from its internal buffer, which contains valid, initialized data (and all of it; that is a- ll of it that had not yet been `consume`d).

With that said, I DO agree that at least because of the last request-response which happens _right before_ the channel starts talking PPP, you might have to switch to a byte-by-byte consumption of the input data, because otherwise you risk "eating up" the beginning of the first incoming PPP packet into the internal buffer of `BufRead` (though I would check if PPP really operates in a way where it might stanrt sending you data immediately after the response to your last AT command).

**OR** - and maybe that's better? - you can just **continue** utilizing the _same_ `BufRead` object that was initially used for AT commands _also_ for the PPP stream? That's completely possible, and BTW layering `Read` over `BufRead` should be trivial I think (though maybe bearing an unnecessary overhead as with PPP you don't need the "naive" overlapping copy, that the hidden `BufRead` would inevitably do).

But... your call. :)

Reply:
I think we will have to create a new `BufReader` in `negotiate()` that is passed into each of the read/write parsing functions.